File: algorithm/src/array/moveZeroes.py
#!/usr/bin/env python3
# -*- codiong:utf-8 -*-

class Solution:
    #def moveZeroes(self, nums: List[int]) -> None:
    def moveZeroes(self, nums):
        """
        Do not return anything, modify nums in-place instead.
        """
        i = 0
        j = 0
        for i in range(len(nums)):
            if 0 == nums[i]:
                i += 1
            else:
                nums[j] = nums[i]
                i += 1
                j += 1
        print(nums)
        while j < len(nums):
            nums[j] = 0
            j += 1
        return nums

def main():
    s = Solution()
    nums = [0,1,0,3,12]
    print(s.moveZeroes(nums))
    

if __name__ == '__main__':
    main()

File: algorithm/src/array/moveZeroes.c
#include "alg_array.h"

/*
 * 暴力
 */
void moveZeroes( int * nums, int numsSize ){
    int i = 0, j = 0, k = 0;
    for( i = 0; i < numsSize - k; i++ ){
        printf( "i,j, k[%d,%d,%d]\n", i, j, k );
        if( 0 == nums[i] ){
            printf( "=i,j, k[%d,%d,%d]\n", i, j, k );
            k++; // 0的个数
            for( j = i; j < numsSize - 1; j++ ){
                nums[j] = nums[j+1];
            }
            i--;   // 连续0的情况
            nums[numsSize - k] = 0; // 发现的0，结尾为0;
        }
        printNums( nums, numsSize );
    }

    /*
    for( j = 0; j < k; j++ ){
        nums[i++] = 0;
    }
    */
}

/*
 * 双指针
 */
void moveZeroes1( int * nums, int numsSize ){
    int i = 0, j = 0, k = numsSize;

    while( i < numsSize ){
        if( 0 == nums[i] ){
            i++;
            k--;
        }else{
            if( i == j ){
                i++;
                j++;
            }else{
                nums[j++] = nums[i];
                nums[i++] = 0;
            }
        }
        printNums( nums, numsSize );
    }
    while( j < numsSize ){
        nums[j++] = 0;
    }
}
